File: ocaml/adarr2.ml
(* Automatic differentiation 
 *
 * Type is now recursive. Trying for higher order derivatives.
 * It will be a miracle if I can get this work.
 *
 * This is hopeless.
 *)


module type ADD = sig
  type t
  val zero : t
  val add : t * t -> t
end

module Unit = struct
  type t = unit
  let zero = ()
  let add ((), ()) = ()
end

module Float = struct
  type t = float
  let zero = 0.0
  let add (x,y) = x +. y
end

module Product (M : ADD) (N: ADD) = struct
  type t = M.t * N.t
  let zero = (M.zero, N.zero)
  let add ((x1,x2), (y1,y2)) = (M.add (x1,y1), N.add (x2,y2))
end

type ('a,'b) darr = DA of (('a -> 'b) * ('a*'b,'a) darr)

module AD (M : ADD) = struct
  (* Cartesian category of differentiable functions *)
  let rec id  = DA (Fun.id, snd)
      and (>>) (DA (f, df) as ff) (DA (g, dg)) = DA (Fun.(f >> g), fst)

      (* and intro x = DA (Fun.intro x, fst) *)
      (* and elim = DA (Fun.elim, fst) *)
      (* and dup = DA (Fun.dup, fst) *)
      (* and add = DA (M.add, fst) *)
      and fst : 'a 'b.('a*'b, 'a) darr = DA (Fun.fst, fst)
      and snd : 'a 'b.('a*'b, 'b) darr = DA (Fun.snd, fst)

      (* and uniti = DA (Fun.uniti, snd >> snd) *)
      (* and unite = DA (Fun.uniti, snd >> uniti) *)
      and swap : 'a 'b.('a*'b, 'b*'a) darr = DA (Fun.swap, snd >> swap)
      (* and first (DA (f,df)) = DA (Fun.first f, fst) *)
      (* and second (DA (f,df)) = DA (Fun.second f, fst) *)
      (* and assocr = DA (Fun.assocr, snd >> assocl) *) 
      (* and assocl = DA (Fun.assocl, snd >> assocr) *)

      (* and (>>) (DA (f, df) as ff) (DA (g, dg)) = DA (Fun.(f >> g), first (dup >> second ff) >> assocr >> second dg >> df) *) 
      (* and intro x = DA (Fun.intro x, fst) *)
      (* and elim = DA (Fun.elim, elim >> intro M.zero) *)
      (* and dup = DA (Fun.dup, snd >> add) *)
      (* and add = DA (M.add, snd >> dup) *)
      (* and fst = DA (Fun.fst, snd >> uniti >> first (intro M.zero) >> swap) *)
      (* and snd = DA (Fun.snd, snd >> uniti >> first (intro M.zero)) *)

      (* and uniti = DA (Fun.uniti, snd >> snd) *)
      (* and unite = DA (Fun.uniti, snd >> uniti) *)
      (* and swap = DA (Fun.swap, snd >> swap) *)
      (* and first (DA (f,df)) = DA (Fun.first f, first fst >> assocl >> first df) *)
      (* and second (DA (f,df)) = DA (Fun.second f, first snd >> swap >> assocr >> second (swap df)) *)
      (* and assocr = DA (Fun.assocr, snd >> assocl) *) 
      (* and assocl = DA (Fun.assocl, snd >> assocr) *)


(*   let rec swap : 'a 'b. (('a * 'b), ('b * 'a)) darr = fun (x,y) -> DR (Utils.Pair.swap (x,y), swap) *)
(*       and unite ((),x) = DR (x, uniti) *)
(*   let rec assocl (x,(y,z))  = DR (((x,y),z), assocr) *)
(*       and assocr ((x,y),z)  = DR ((x,(y,z)), assocl) *)

(*   let rec inl x = DR (Left x, merge) *)
(*       and inr x = DR (Right x, merge) *)
(*       and merge = function *) 
(*                   | Left x -> DR (x, inl) *) 
(*                   | Right y -> DR (y, inr) *)

(*   let rec mirror x = DR (Utils.mirror x, mirror) *)
(*   let rec dup x     = DR ((x,x), add) *)
(*       and add xy    = DR (M.add xy, dup) *)

(*   let ( && ) f g x = (dup >> (f ** g)) x *)
(*   let snd x = (first elim >> unite) x *)
(*   let fst x = (swap >> snd) x *)

(*   let rec ( >> ) : 'a 'b 'c. ('a -> ('a,'b) dres) -> ('b -> ('b,'c) dres) -> 'a -> ('a,'c) dres = fun f g x -> *)
(*     let DR (y,df) = f x in *)
(*     let DR (z,dg) = g y in *)
(*     DR (z, fun dz -> (dg >> df) dz) *)

(*   let rec left f = function *) 
(*     | Left x -> let DR (z,df) = f x in DR (Left z, left df) *)
(*     | Right y -> DR (Right y, id) *)

(*   let rec first : 'a 'b 'c. ('a,'b) darr -> ('a * 'c, 'b * 'c) darr = fun f (x,y) -> *)
(*     let DR (z, df) = f x in *)
(*     DR ((z,y), first df) *)

(*   let second f x = (swap >> first f >> swap) x *)
(*   let right f x = (mirror >> left f >> mirror) x *)
(*   let ( ** ) f g x = (first f >> second g) x *)
(*   let ( ++ ) f g x = (left f >> right g) x *)
(*   (1* let ( || ) f g x = ((f ++ g) >> merge) x *1) *)

(*   let rec neg x     = DR (~-. x, neg) *)
(*   let rec scale k x = DR (k *. x, scale k) *)
(*   let log x = DR (Pervasives.log x, scale (1.0 /. x)) *)
(*   let exp x = let y = Pervasives.exp x in DR (y, scale y) *)

(*   let grad f x = *) 
(*     let DR (y,df) = f x in *) 
(*     let DR (dy_dx,_) = df 1.0 in *)
(*     (y, dy_dx) *)
  
   (* let grad (DA (f,df)) = uniti >> first (intro 1.0) >> swap >> df *)
   let run (DA (f,_)) x = f x
   let ( |>> ) x f = run f x 
end

module ADF = AD (Float)
(* module ADP = HAD (Product (Float) (Float)) *)
(* module ADU = HAD (Unit) *)
open ADF

(* let mul (x,y)  = DR HAD.(x *. y, ADF.(scale y && scale x)) *)

File: haskell/adarr2.hs
{-# LANGUAGE TypeOperators, FlexibleInstances #-}
import Prelude hiding ((>>), (**), (&&), id, fst, snd, exp, log)
import qualified Prelude

-- This is an experiment in functional reverse mode automatic differentiation using
-- a sort of arrow-ish system to represent a dataflow graph that includes the computation
-- of derivatives as well as the computation a primal function.
--
-- It doesn't exactly fit the Arrow type class, partly because there is no way to lift
-- an arbitrary function to a differentiable function and thus no way to implement arr,
-- and also because the differentiability requirement means that most of the data types
-- flowing through the graph need to be able to represent their own tangent spaces; that is,
-- we need `add` and `zero` operators for every data type.
--
-- Instead, the space of differentiable functions is a closer fit to a Cartesian or 
-- bi-Cartesian category. Below, I've only implemented a Cartesian category that can
-- compute with product types. The set of primitives is modelled on Roshan and Sabry's
-- work on reversible computing and information effects [1,2]. The back-propagation of
-- derivatives is not completely unlike reverse computation, and also idea of information
-- effects inherent in certain operations that create or destroy information seems to
-- have resonance in here in the way that dup, intro and elim work. Conal Elliot's work
-- on compiling functions to circuits in closed bi-Cartesian categories was also an influence.
--
-- I orignally wanted to define a class CartesianCategory with instances for ordinary
-- functions (->) and for differentiable functions (~>), but this turned out to be rather
-- difficult because of the additional constraints on types imposed by the differentiable
-- version. The type class Plus states that a type forms an additive monoid and is needed
-- to add derivatives or return an appropriate zero. The differentiable category needs a lot
-- of Plus type constraints which seem to disqualify it from being a general cartesian category.
--
-- So instead, I define the functions for the (->) category with primed names (like id')
-- and the functions for the (~>) with unprimed names, hiding the corresponding ones from
-- Prelude where necesssary.
--
-- [1] Roshan P James and Amr Sabry. Information effects. 
--     In ACM SIGPLAN Notices, volume 47, pages 73–84. ACM, 2012.
-- [2] Roshan P James and Amr Sabry. The two dualities of computation: 
--     Negative and fractional types. Technical report, Citeseer, 2012. 

infixr 0 ~>
infixr 7 **
infixr 7 &&
infixr 0 |>

class Plus t where
	plus :: (t,t) -> t 
	zero :: t 

instance (Num t) => Plus t where
	plus (x,y) = x + y
	zero = 0

instance Plus () where
	plus ((),()) = ()
	zero = ()

instance (Plus t, Plus u) => Plus (t,u) where
  plus ((x1,x2), (y1,y2)) = (plus (x1,y1), plus (x2,y2))
  zero = (zero,zero)

-- class CartesianCategory k where
--   id     ::  a `k` a
--   (>>)   :: (a `k` b) -> (b `k` c) -> (a `k` c)

--   -- wiring (type isomorphisms)
--   uniti  :: a `k` ((),a)
--   unite  :: ((),a) `k` a
--   swap   :: (a,b) `k` (b,a)
--   assocl :: (a,(b,c)) `k` ((a,b),c)
--   assocr :: ((a,b),c) `k` (a,(b,c))

--   -- information effects (see Roshan and Sabry)
--   intro  :: a -> () `k` a
--   elim   :: a `k` ()
--   dup    :: a `k` (a,a)
--   fst    :: (a,b) `k` a
--   snd    :: (a,b) `k` b
--
--   -- Arrow combinators
--   first  :: (a `k` b) -> (a,c) `k` (b,c)
--   second :: (a `k` b) -> (c,a) `k` (c,b)
--   (**) :: (a `k` b) -> (c `k` d) -> (a,c) `k` (b,d)
--   (&&) :: (a `k` b) -> (a `k` c) -> a `k` (b,c)


-- instance CartesianCategory (->) where
--    ... mostly obvious - see below ...

id' x = x
intro' x () = x
elim' _ = ()
dup' x = (x,x)
uniti' x = ((),x)
unite' (_,x) = x
swap' (x,y) = (y,x)
fst' (x,_) = x
snd' (_,y) = y
assocl' (x,(y,z)) = ((x,y),z)
assocr' ((x,y),z) = (x,(y,z))
first' f (x,y) = (f x, y)
second' f (x,y) = (x, f y)

data a ~> b = AD (a->b) ((b,a) ~> a)

-- instance CartesianCategory (~>) where
--    ... won't fly due to Plus type constraints everywhere

id ::  Plus a => a ~> a
id = AD id' fst

(>>) :: (Plus a, Plus b, Plus c) => (a ~> b) -> (b ~> c) -> (a ~> c)
ff@(AD f df) >> AD g dg = AD (g . f) (second (dup >> first ff) >> assocl >> first dg >> df)

intro x = AD (intro' x) fst
elim :: Plus t => t ~> ()
elim  = AD elim' (fst >> intro zero)

dup :: Plus t => t ~> (t,t)
dup = AD dup' (fst >> add)

add :: Plus t => (t,t) ~> t
add = AD plus (fst >> dup)

uniti :: Plus t => t ~> ((),t)
uniti = let fs = fst >> swap in AD uniti' (fs >> fs)

unite :: Plus t => ((),t) ~> t
unite = AD unite' (fst >> uniti)

fst :: (Plus a, Plus b) => (a,b) ~> a
fst = AD fst' (fst >> uniti >> swap >> second (intro zero))

snd :: (Plus a, Plus b) => (a,b) ~> b
snd = AD snd' (fst >> uniti >> first (intro zero))

swap :: (Plus a, Plus b) => (a,b) ~> (b,a)
swap = AD swap' (fst >> swap)

assocl :: (Plus a, Plus b, Plus c) => (a,(b,c)) ~> ((a,b),c)
assocl = AD assocl' (fst >> assocr)

assocr :: (Plus a, Plus b, Plus c) => ((a,b),c) ~> (a,(b,c))
assocr = AD assocr' (fst >> assocl)

first :: (Plus a, Plus b, Plus c) => (a ~> b) -> (a,c) ~> (b,c)
first (AD f df) = AD (first' f) (second fst >> swap >> assocl >> first (swap df))

second :: (Plus a, Plus b, Plus c) => (a ~> b) -> (c,a) ~> (c,b)
second (AD f df) = AD (second' f) (second snd >> assocr >> second df)

f ** g = first f >> second g
f && g = dup >> (f ** g)

mul :: (Num t, Plus t) => (t,t) ~> t
mul = AD (uncurry (Prelude.*)) 
         (second dup 
          >> assocl 
          >> first (assocr >> swap >> first mul) 
          >> assocr >> second mul)

scale :: (Num t, Plus t) => t -> t ~> t
scale k = AD (Prelude.* k) (snd >> scale k)

neg :: (Num t, Plus t) => t ~> t
neg = scale (-1)

pow 0 = elim >> intro 1
pow 1 = id
pow k = AD (Prelude.** k) (first (pow (k-1) >> scale k) >> mul)

exp :: (Floating t, Plus t) => t ~> t
exp = AD Prelude.exp (first exp >> mul) 

log :: (Eq t, Floating t, Plus t) => t ~> t
log = AD Prelude.log (first (pow (-1)) >> mul)

grad (AD f df) = uniti >> first (intro 1.0) >> swap >> df

x |> AD f df = f x
